[JSC] Optimize Promise.all

Constellation opened the pull request:
#### 4136973865587fb6adf30abb4ca51fdfae868aed
<pre>
[JSC] Optimize Promise.all
<a href="https://bugs.webkit.org/show_bug.cgi?id=244239">https://bugs.webkit.org/show_bug.cgi?id=244239</a>

Reviewed by Ross Kirsling.

This patch optimizes Promise.all. Previously we are always creating a promise per element.
Also, we are allocating closure which resolves promise returned by Promise.all.

1. By using promise resolution context mechanism, we no longer allocate closures.
2. By checking returned promise from Promise.resolve carefully, we avoid creating a promise,
   which is not visible to users.

We observe performance improvement in promise microbenchmarks.

    ToT
    Time(parallel-async-es2017-native): 42.6 ms.
    Time(parallel-promises-es2015-native): 32.1 ms.

    Patched
    Time(parallel-async-es2017-native): 32.9 ms.
    Time(parallel-promises-es2015-native): 23.1 ms.

* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(linkTimeConstant.promiseOnRejectedWithContext):
(linkTimeConstant.promiseAllOnFulfilled):
(linkTimeConstant.promiseNewOnRejected.return.reject):
(linkTimeConstant.promiseNewOnRejected):
(linkTimeConstant.promiseAllNewResolveElement):
(all):
(all.newResolveElement): Deleted.

Canonical link: <a href="https://commits.webkit.org/253716@main">https://commits.webkit.org/253716@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/88be21bfd86dddfeb1632959f4bd3368fd0f63e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30966 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17758 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95712 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149470 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29328 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79031 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90938 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23679 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23705 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/78962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66710 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78795 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27080 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72452 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27014 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/13829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25839 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2635 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28693 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75237 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28637 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/33108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16641 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -78,6 +76,109 @@ function all(iterable)
+    };
+}
+
+function all(iterable)
```
is a similar optimization relevant for `InternalPromise`? Could be good for ESM loader perf

Constellation replied:
Because `Promise.all` can have many user-observable behavior, we should not use it for internal promise, let's use `@InternalPromise.internalAll` instead :)

Constellation commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -78,6 +79,106 @@ function all(iterable)
+                    var promiseOrCapability = @newPromiseCapabilitySlow(constructor);
+                @performPromiseThen(nextPromise, @promiseAllOnFulfilled, @promiseOnRejectedWithContext, promiseOrCapability, { globalContext, index });
+            } else {
+                if (!onRejected)
+                    onRejected = @promiseNewOnRejected(promise);
+                then.@call(nextPromise, @promiseAllNewResolveElement(globalContext, index), onRejected);
```
We need to create these closures for handlers newly, since this `then` can be anything (it is not default then), so passed functions are user-observable.

rkirsling commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -78,6 +79,106 @@ function all(iterable)
+        var values = globalContext.values;
+        @putByValDirect(values, index, argument);
+
+        if (!--globalContext.remainingElementsCount)
+            return @resolvePromiseWithFirstResolvingFunctionCallCheck(globalContext.promise, values);
```
nit: Seems like we could keep the old and new style in alignment with each other (wrt to `!--` here and final `return @undefined` there).

Constellation replied:
We changed to make using `return @resolvePromiseWithFirstResolvingFunctionCallCheck` thing in the other code too.

rkirsling commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -78,6 +79,106 @@ function all(iterable)
+            if (@isPromise(nextPromise) && then === @defaultPromiseThen) {
+                var constructor = @speciesConstructor(nextPromise, @Promise);
+                if (constructor !== @Promise)
+                    var promiseOrCapability = @newPromiseCapabilitySlow(constructor);
```
I know we have to use `var`s in builtins JS, but I definitely don't think we should hoist them. 😅

Constellation replied:
Added `var promiseOrCapability` thing.

rkirsling commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -78,6 +79,106 @@ function all(iterable)
 }
 
+@linkTimeConstant
+function promiseOnRejectedWithContext(argument, context)
```
Could we name this `promiseAllOnRejected` for consistency? It looks quite generic at a glance, but in actuality `context.globalContext.promise` is unlikely to occur elsewhere, right?

Constellation replied:
I'm now intentionally naming this to make it usable for the other promise constructor functions (`Promise.race`, `Promise.any`, `Promise.allSettled`), but right now it is not used yet.

rkirsling reviewed (approved):
r=me, aside from the hoisting. 😛

Constellation committed:
[JSC] Optimize Promise.all
https://bugs.webkit.org/show_bug.cgi?id=244239

Reviewed by Ross Kirsling.

This patch optimizes Promise.all. Previously we are always creating a promise per element.
Also, we are allocating closure which resolves promise returned by Promise.all.

1. By using promise resolution context mechanism, we no longer allocate closures.
2. By checking returned promise from Promise.resolve carefully, we avoid creating a promise,
   which is not visible to users.

We observe performance improvement in promise microbenchmarks.

    ToT
    Time(parallel-async-es2017-native): 42.6 ms.
    Time(parallel-promises-es2015-native): 32.1 ms.

    Patched
    Time(parallel-async-es2017-native): 32.9 ms.
    Time(parallel-promises-es2015-native): 23.1 ms.

* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(linkTimeConstant.promiseOnRejectedWithContext):
(linkTimeConstant.promiseAllOnFulfilled):
(linkTimeConstant.promiseNewOnRejected.return.reject):
(linkTimeConstant.promiseNewOnRejected):
(linkTimeConstant.promiseAllNewResolveElement):
(all):
(all.newResolveElement): Deleted.

Canonical link: https://commits.webkit.org/253716@main
